compiler/src/codegen/garbage_collection.re

```diff
@@ -0,0 +1,781 @@
+        MReturnCallIndirect({
+          ...data,
+          func: handle_imm(~is_return=true, func),
+          args: List.map(handle_imm(~is_return=true), args),
+        })
+      | MError(e, args) => MError(e, List.map(handle_imm, args))
+      | MAllocate(alloc) =>
+        let alloc =
+          switch (alloc) {
+          | MClosure({variables} as cdata) =>
+            MClosure({...cdata, variables: List.map(handle_imm, variables)})
+          | MTuple(args) => MTuple(List.map(handle_imm, args))
+          | MBox(imm) => MBox(handle_imm(imm))
+          | MArray(args) => MArray(List.map(handle_imm, args))
+          | MRecord(tag, fields) =>
+            MRecord(
+              handle_imm(tag),
+              List.map(
+                ((field, value)) => (field, handle_imm(value)),
+                fields,
+              ),
+            )
+          | MADT(ttag, vtag, args) =>
+            MADT(
+              handle_imm(ttag),
+              handle_imm(vtag),
+              List.map(handle_imm, args),
+            )
+          | MBytes(_)
+          | MString(_)
+          | MInt32(_)
+          | MInt64(_)
+          | MUint32(_)
+          | MUint64(_)
+          | MFloat32(_)
+          | MFloat64(_)
+          | MRational(_)
+          | MBigInt(_) => alloc
+          };
+        MAllocate(alloc);
+      | MTagOp(tag_op, tag_type, imm) =>
+        MTagOp(tag_op, tag_type, handle_imm(~non_gc_instr=true, imm))
+      | MArityOp(arity_operand, arity_op, imm) =>
+        MArityOp(
+          arity_operand,
+          arity_op,
+          handle_imm(~non_gc_instr=true, imm),
+        )
+      | MIf(imm, thn, els) =>
+        MIf(
+          handle_imm(~non_gc_instr=true, imm),
+          apply_gc(thn),
+          apply_gc(els),
+        )
+      | MFor(cond, inc, body) =>
+        MFor(
+          Option.map(apply_gc, cond),
+          Option.map(apply_gc, inc),
+          apply_gc(~loop_context=[level, ...loop_context], body),
+        )
+      | MContinue
+      | MBreak => instr.instr_desc
+      | MReturn(Some({instr_desc: MImmediate(imm)} as instr)) =>
+        // A return will use the argument directly without the need to incref
+        MReturn(
+          Some({
+            ...instr,
+            instr_desc: MImmediate(handle_imm(~is_return=true, imm)),
+          }),
+        )
+      | MReturn(ret) => MReturn(Option.map(process_instr, ret))
+      | MSwitch(value, branches, default, alloc) =>
+        MSwitch(
+          handle_imm(value),
+          List.map(
+            ((branch_id, branch)) => (branch_id, apply_gc(branch)),
+            branches,
+          ),
+          apply_gc(default),
+          alloc,
+        )
+      | MPrim0(_) => instr.instr_desc
+      | MPrim1(WasmFromGrain, imm) =>
+        // special case: we delay cleanup of any `fromGrain` to the end of the function to
+        // remove the headache of the value being GC'd before anything can be done with it
```

Comment:
Not that it should be done in this PR but what other options were considered for making `fromGrain` not a special case?

Reply:
The only other option would be to have us _always_ use `ignore` to hold onto the reference. We could maybe write some sort of well-formedness check to try to catch when we let something die too soon, but it's really tricky.